### PR TITLE
[Bugfix] Remove X-Request-Id format restrictions and make size configurable

### LINKED ISSUE #20688: [Feature Request]  Make X-Request-Id validation configurable
### Is your feature request related to a problem? Please describe

Hi OpenSearch Team,

I’m looking at PR #19798 where `X-Request-Id` is currently validated as **hex-only** and **exactly 32 chars**.

* Many gateways / load balancers emit request IDs that are not hex-only.
* Operators often want to pass through an existing correlation ID end-to-end without rewriting.

Before I start implementing anything, I’d like to ask if you would be OK with me opening a PR that removes the **hex-only** restriction and makes **only the length configurable**, while keeping the current behavior as the default.

### Describe the solution you'd like

* Introduce a setting like `http.request_id.max_length` (default: **32**).
* Validation would enforce:

  * header is present / non-empty (as today)
  * length is `<= max_length` (default keeps 32)
* The **hex character check would be removed**, so operators can propagate common request-id formats (UUID/ULID/base64url-like tokens, vendor-specific IDs) unchanged.

### Related component

Search

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by finnegancarroll**
### Description
Introduces dynamic `http.request_id.max_length` setting for configuring maximum length for X-Request-Id headers. Removes the alpha-numeric validation of X-Request-Id.

### Related Issues
Resolves #20688

### Check List
- [x] Functionality includes testing.
- [x] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by finnegancarroll** [comment]
Code analyzer gives the following concern.
```
allows arbitrary characters (including newlines, control chars, or special sequences) in the request ID, which could enable log injection 
```

Gave this a quick test and it seems netty rejects these headers as invalid so we do not need to handle them ourselves.
```
curl -s --http1.1 -H $'X-Request-Id: foo\rbar' \
  -X GET "$BASE/test-index/_search" \
  -H "Content-Type: application/json" \
  -d '{"query": {"match_all": {}}}' | jq .
```
```
{
  "error": {
    "root_cause": [
      {
        "type": "illegal_argument_exception",
        "reason": "Validation failed for header 'X-Request-Id'"
      }
    ],
    "type": "illegal_argument_exception",
    "reason": "Validation failed for header 'X-Request-Id'",
    "caused_by": {
      "type": "illegal_argument_exception",
      "reason": "a header value contains prohibited character 0xd at index 3."
    }
  },
  "status": 400
}
```

**Comment by finnegancarroll** [comment]
@sgup432 can you take a look when you have a chance?

File: server/src/main/java/org/opensearch/http/HttpTransportSettings.java [hunk 1]
```diff
@@ -259,6 +259,14 @@ public final class HttpTransportSettings {
         Setting.Property.NodeScope
     );
 
+    public static final Setting<Integer> SETTING_HTTP_REQUEST_ID_MAX_LENGTH = intSetting(
+        "http.request_id.max_length",
+        128,
+        1,
```
Comment (sgup432) [inline server/src/main/java/org/opensearch/http/HttpTransportSettings.java]: Do you think we should increase this min value to something like 16 atleast? As we shouldn't allow someone passing a header id with size as 1.

File: server/src/test/java/org/opensearch/common/util/RequestUtilsTests.java [hunk 2]
```diff
@@ -13,68 +13,94 @@
 
 public class RequestUtilsTests extends OpenSearchTestCase {
```
Comment (sgup432) [inline server/src/test/java/org/opensearch/common/util/RequestUtilsTests.java:16]: Can we also add some sanity IT tests(apart from this unit tests) around updating the requestId max length dynamically? And test both positive and negative flow. One basic should suffice.

File: server/src/test/java/org/opensearch/common/util/RequestUtilsTests.java [hunk 3]
```diff
@@ -13,68 +13,94 @@
...
 
+    private static final int DEFAULT_MAX_LENGTH = 128;
```
Comment (sgup432) [inline server/src/test/java/org/opensearch/common/util/RequestUtilsTests.java]: I would prefer to fetch this value from the settings itself instead of hardcoding.

File: server/src/main/java/org/opensearch/http/HttpTransportSettings.java (hunk 1 above)
Comment (sgup432) [reply server/src/main/java/org/opensearch/http/HttpTransportSettings.java]: In addition, we should also add a max value here? To avoid issues where a user can pass a very big header erroneously or intentionally?

Comment (finnegancarroll) [reply server/src/main/java/org/opensearch/http/HttpTransportSettings.java]: Added a `1024` max to guard against excessively large ids. wdyt about keeping the min at 1? I think enforcing some other min value implies a range for valid id length instead of strict maximum and I can't think of any harm here.

File: server/src/test/java/org/opensearch/common/util/RequestUtilsTests.java (hunk 2 above)
Comment (finnegancarroll) [reply server/src/test/java/org/opensearch/common/util/RequestUtilsTests.java:16]: Since this change makes `validateRequestId` trivial I ended up migrating the remaining UTs into integ tests as well.

File: server/src/main/java/org/opensearch/http/HttpTransportSettings.java (hunk 1 above)
Comment (sgup432) [reply server/src/main/java/org/opensearch/http/HttpTransportSettings.java]: Yeah having some min value does help here. I am not sure what is the right value to start with, maybe 16?

File: server/src/test/java/org/opensearch/common/util/RequestUtilsTests.java [hunk 4]
```diff
@@ -1,80 +0,0 @@
...
- * this file be licensed under the Apache-2.0 license or a
- * compatible open source license.
- */
-
-package org.opensearch.common.util;
-
-import org.opensearch.core.common.Strings;
-import org.opensearch.test.OpenSearchTestCase;
-
-public class RequestUtilsTests extends OpenSearchTestCase {
```
Comment (sgup432) [inline server/src/test/java/org/opensearch/common/util/RequestUtilsTests.java:14]: Oh I didn't mean to remove this unit test. We still need this one.

I just asked to write a simple dynamic setting IT test(in an appropriate IT class) to test for the flow: 
1. Trigger search request with a request_id length like 64 which is <=max 
2. Update the setting to set the max value to 32 dynamically. 
3. Trigger the search request again, this time it should fail with invalid request_id.

Comment (finnegancarroll) [reply server/src/test/java/org/opensearch/common/util/RequestUtilsTests.java:14]: Got it. Added back this UT suite and added IT with the above flow.

File: server/src/main/java/org/opensearch/http/HttpTransportSettings.java (hunk 1 above)
Comment (finnegancarroll) [reply server/src/main/java/org/opensearch/http/HttpTransportSettings.java]: Moved this setting to 16 minimum.

**Review by bowenlan-amzn — APPROVED** [review]
Looks good! 
Regarding potential character injection concerns — I guess an attacker would need existing access and permissions to the cluster to exploit this, at which point they have far more impactful ways to cause damage.

**Review by jainankitk — APPROVED** [review]
Looks good to me!

**Review by sgup432 — APPROVED** [review]

**Review by andrross — APPROVED** [review]
